### ISSUE TITLE
[BUG] lifecycle callbacks called in wrong sequence for top level route.

### ISSUE THREAD

**Issue by huochunpeng** [aurelia/router#612]
**I'm submitting a bug report**

* **Library Version:**
all
It is a regression caused by aurelia-templating v1.8.2.
But I am pretty sure this is a router bug (whether the buggy code is in this repo, I am not sure), the new aurelia-templating fix just made this router bug appears even worse.

* **Operating System:**
OSX 10.13

* **Node Version:**
8.11.3

* **NPM Version:**
6.2.0

* **JSPM OR Webpack AND Version**
NA

* **Browser:**
all

* **Language:**
all

**Current behavior:**
https://github.com/huochunpeng/router-bug2

app.js has one level router config with two pages (page-one.js and page-two.js), here is the log of lifecycle callbacks

with aurelia-templating 1.8.1 (with or without bluebird)
note bind() is in right order, but attached() is not (page-one should be attached before app)

```
[Log] app configureRouter (app-bundle.js, line 22)
[Log] app bind (app-bundle.js, line 34)
[Log] page-one bind (app-bundle.js, line 105)
[Log] app attached (app-bundle.js, line 38)
[Log] page-one attached (app-bundle.js, line 109)
```

with aurelia-templating 1.8.2 (with or without bluebird)
bind() is in wrong order, plus wrong attached() order

```
[Log] app configureRouter (app-bundle.js, line 22)
[Log] page-one bind (app-bundle.js, line 105)
[Log] app bind (app-bundle.js, line 34)
[Log] app attached (app-bundle.js, line 38)
[Log] page-one attached (app-bundle.js, line 109)
```

Due to the wrong bind() call sequence, the page-one now doesn't get correct inherit overrideContext from app.

User has reported missing variable (defined in top app.js) in router loaded component view template.
https://discourse.aurelia.io/t/using-require-breaks-router/1411/31?u=huochunpeng

This bug is related to #548, which can be fixed by #571. But that fix only fix lifecycle callbacks for nested route, not the top route (as top route is not delayed).

**Expected/desired behavior:**

The correct sequence should be:
```
app configureRouter
app bind
page-one bind
page-one attached
app attached
```

No matter what aurelia-templating version in use, our sequence is NOT right.
aurelia-templating v1.8.2 made this bug even worse, and more noticeable to end user.

* **What is the expected behavior?**

* **What is the motivation / use case for changing the behavior?**

The user experienced missing variable (due to missing overrideContext) is not the bug we want to fix.
The wrong lifecycle callback sequence is the bug to be solved.

cc @bigopon @davismj @EisenbergEffect 

aurelia-templating v1.8.2 just killed off everyone (intentionally or unintentionally) rely on the default inheritBindingContext behaviour of router loaded component.

**Comment by bigopon** [comment]
> note bind() is in right order, but attached() is not (page-one should be attached before app)

This is not correct. Child element cannot have `attached` called before parent's `attached` gets called.

**Comment by huochunpeng** [comment]
@bigopon right, so with 1.8.1, the behaviour is expected.

I just tested by removing router from the picture.

app.html
```html
<template>
  <require from="./page-one"></require>
  <page-one></page-one>
</template>
```

Both 1.8.1 and 1.8.2 behaves same:
```
app bind
page-one bind
app attached
page-one attached
```

So this is likely still a router related bug.

**Comment by bigopon** [comment]
It's likely to be in `templating-router` + `templating` than router I think. 

https://github.com/aurelia/templating-router/blob/e83bd1bc182350c7cdc66736404633a05cbb711b/src/router-view.js#L131-L136

It's a bit early compared to what we have in `templating@1.8.2`

**Comment by huochunpeng** [comment]
Means an easy fix for you?

**Comment by bigopon** [comment]
> and secondly that process happens before bind.

Isn't this the expected behavior ? A view port (read `<router-view/>` element) should process (nav, route, router) before it bind the view and view model together

**Comment by StrahilKazlachev** [comment]
Does anything guarantee that the `<router-view>` itself is bound when [this line](https://github.com/aurelia/templating-router/blob/a465e6e43c08533e7c9e45f557581fba1e11bcdc/src/router-view.js#L115) is hit?

**Comment by bigopon** [comment]
I haven't had any immediately obvious fix for this atm, so the suggestion is to undo https://github.com/aurelia/templating/pull/633 via following block

```js
import { CompositionEngine } from 'aurelia-templating';

CompositionEngine.prototype._createControllerAndSwap = function(context) {
    return this.createController(context).then(controller => {
      controller.automate(context.overrideContext, context.owningView);

      if (context.compositionTransactionOwnershipToken) {
        return context
          .compositionTransactionOwnershipToken
          .waitForCompositionComplete()
          .then(() => {
            return this._swap(context, controller.view);
          })
          .then(() => controller);
      }

      return this._swap(context, controller.view).then(() => controller);
    });
  }
```

and avoid using `bluebird` to work around the original issue.

**Comment by StrahilKazlachev** [comment]
@bigopon why? Where is the code path that calls `.compose`, since I don't see how else you can reach `_createControllerAndSwap`? I can't find any code in `aurelia-router` that uses the `CompositionEngine`. `<router-view>` uses `.createController`, from there you can't reach `_createControllerAndSwap`, and then itself "automates" the `Controller`s.
I'm not so familiar with the router system so can you look at [this](https://github.com/StrahilKazlachev/templating-router/tree/fix/router-view/bind-automate-race). I'm unsure if this won't lock up the router and the application so ...

**Comment by bigopon** [comment]
The fix at https://github.com/aurelia/templating/pull/633 pushes app root `bind` to happen after any `<router-view/>` view model `bind()`, which is a not nice change. Undoing that PR brings back the old behavior, which I think is a better temporary solution.

The way the root router & child router work, though may be not nice (`bind` happens inside out), but it's been there, & usable.

---

For https://github.com/StrahilKazlachev/templating-router/commit/c54c70697fea79e2d28b545cd7f4ca4b36d28089, Very nice solution, it also fixes the issue with `bind` happens inside out. A potential tweak is that depends on whether `<router-view/>` inside an `if` (or any other non `repeat` template controller with caching behavior) is a supported scenario, we may need to tweak it a little bit.

What the behavior is like with @StrahilKazlachev 's solution

![lc](https://user-images.githubusercontent.com/9994529/43366363-46cad304-9380-11e8-9cb6-eea55441d517.png)


Though I cannot get it to work with child router without bringing back behavior of `1.8.1`, did it work for you @StrahilKazlachev  ?

Though I'm not confident we can fix the inside out behavior without a version bump.

cc @EisenbergEffect @fkleuver @davismj @huochunpeng

**Comment by StrahilKazlachev** [comment]
@bigopon my *"solution"* did lock the app startup with `1.8.2`. Basically the `.bind` never gets called, the `Promise` never gets resolved.

**Comment by huochunpeng** [comment]
@bigopon for nested child routes, if you bring #571, it might just works.

**Comment by jvlobo** [comment]
Hi there!!

I've just opened this [Stackoverflow question](https://stackoverflow.com/questions/52435440/main-component-publishing-event-before-child-is-loaded) and seem like the problem is the one you guys are discussing here.

Any news on it?

Cheers :)

**Comment by huochunpeng** [comment]
@jvlobo unfortunately the team has decided to live with this bug at the moment. This is one of domino bugs, created by a chain of patches to fix some other nasty bugs. Comparing to other bugs, this one is the least serious one. We are bit lazy here, or should I say bit scared to create another new bug with another patch :-)

Don't worry, vNext will cleanup all of them :D (@fkleuver sweating heavily when reading this)

**Comment by jvlobo** [comment]
I've been having a lot of bugs recently with some of the updates :(
I'll try to do a workaround for this one too, since our app is in production.
Thank you.

**Comment by huochunpeng** [comment]
@jvlobo sorry to hear that. Could you post some code to discourse forum so folks can help?

**Comment by jvlobo** [comment]
Where is the discourse forum? Thanks!

**Comment by huochunpeng** [comment]
Man, on aurelia home page :-) https://discourse.aurelia.io

**Comment by fkleuver** [comment]
@huochunpeng  vNext won't need to clean any of this up since the new core framework will make it much simpler to achieve the same things. Timing related issues will be pretty much nonexistent. I'm not sweating at all :)
